### PR TITLE
State vectors iterator

### PR DESCRIPTION
**Pull request by jmbarr**
Does as suggested in #673. Iterating over `StateVectors` now yields a `StateVector`. Note that this breaks a few tests (which have been amended), so may also break iterations which previously accommodated the old way. 

Also fails doc tests. No idea why that is and doesn't appear to be to do with these changes.

### CODE REVIEW

**Comment by sdhiscocks** [comment]
> Also fails doc tests. No idea why that is and doesn't appear to be to do with these changes.

Looks like these changes cause the _Moving Platform Simulator_ example to hang.

File: stonesoup/types/array.py [hunk 1]
```diff
@@ -180,14 +185,15 @@ def _average(state_vectors, axis=None, weights=None, returned=False):
         elif axis == 1:  # Need to handle special cases of averaging potentially
             state_vector = StateVector(
                 np.empty((state_vectors.shape[0], 1), dtype=state_vectors.dtype))
-            for dim, row in enumerate(state_vectors):
+            for dim, row in enumerate(state_vectors.T):
```
Comment (sdhiscocks) [inline stonesoup/types/array.py:188]:
```suggestion
            for dim, row in enumerate(np.asarray(state_vectors)):
```

File: stonesoup/types/array.py [hunk 2]
```diff
@@ -180,14 +185,15 @@ def _average(state_vectors, axis=None, weights=None, returned=False):
...
                 type_ = type(row[0])  # Assume all the same type
                 if hasattr(type_, 'average'):
                     # Check if type has custom average method
                     state_vector[dim, 0] = type_.average(row, weights=weights)
                 else:
                     # Else use numpy built in, converting to float array
-                    state_vector[dim, 0] = type_(np.average(np.asfarray(row), weights=weights))
+                    state_vector[dim, 0] = type_(np.average(np.asfarray(row).flatten(),
+                                                            weights=weights))
```
Comment (sdhiscocks) [inline stonesoup/types/array.py:196]:
```suggestion
                    state_vector[dim, 0] = type_(np.average(row, weights=weights))
```

**Review by jswright-dstl — APPROVED** [review]

**Review by nperree-dstl — APPROVED** [review]

File: stonesoup/models/measurement/tests/test_combined.py [hunk 3]
```diff
@@ -72,10 +72,10 @@ def test_rvs(model):
...
     rvs_state = model.rvs(10)
     assert rvs_state.shape == (4, 10)
-    assert all(isinstance(state, Bearing) for state in rvs_state[0])
-    assert all(not isinstance(state, Bearing) for state in rvs_state[1])
-    assert all(isinstance(state, Bearing) for state in rvs_state[2])
-    assert all(not isinstance(state, Bearing) for state in rvs_state[3])
+    assert all(isinstance(state[0], Bearing) for state in rvs_state[0])
+    assert all(not isinstance(state[0], Bearing) for state in rvs_state[1])
+    assert all(isinstance(state[0], Bearing) for state in rvs_state[2])
+    assert all(not isinstance(state[0], Bearing) for state in rvs_state[3])
```
Comment (sdhiscocks) [inline stonesoup/models/measurement/tests/test_combined.py:78]: I was a little confused by this change, but now looking more closely, as I can see it's an issue caused by ending up with `StateVectors` type when slicing in certain ways. The new iteration method has highlight this bug here.

I'll push commit here with fix for that, which probably will also resolve some other issues.

**Pull request merged**